コメントを付けられるようにする

yumiya-g opened the pull request:
サンプルアプリにコメント投稿機能を実装しました。

yumiya-g committed:
Add: ReportモデルのCRUDをscaffoldで作成しました

- JSONレスポンス用のファイルは削除

yumiya-g committed:
Fix: データベースをリセットして、マイグレーションを実行しました

yumiya-g committed:
Add: 日報(report)モデルに、user_idカラムを追加しました

- 各日報に作成者情報を持たせるため

yumiya-g committed:
Fix: 日報作成者のみ、編集と削除を実行できるよう調整しました

- 日報のデータベースに、作成者のID（`user_id`）を追加する
- ログインユーザーと日報作成者のIDが一致しなければ、編集・削除を実行させず、`show`アクションへリダイレクト

yumiya-g committed:
Add: コメント（comment）モデルを作成して、他のモデルとポリモーフィック関連付けを行いました

yumiya-g committed:
Add: 本（book）モデルと日報（report）モデルの入れ子に、コメント（comment）モデルを設定しました

yumiya-g committed:
Fix: コメントモデルが単数形になっていたので複数型（comments）に修正しました

yumiya-g committed:
Add: 日報モデルにコメント機能を追加しました

- コメントフォームは部分テンプレートにして、使い回す
- コメントしたユーザーを削除したら、ユーザ名は「退会済みユーザー」とする

yumiya-g committed:
Add: Commentモデルにバリデーションを追加しました

- コメントが空（empty）でないことを確認する

yumiya-g committed:
Fix: タイムゾーンを日本時間（Tokyo）に変更しました

yumiya-g committed:
Fix: User,Report,Commentモデルの関連付けを修正しました

yumiya-g committed:
Add: 投稿時刻を日本語で表示するキー`ja_short`を追加しました

yumiya-g committed:
Add: ヘッダーメニューに「日報」一覧のリンクを追加しました

yumiya-g committed:
Fix: 日報に投稿されたコメントを取得する処理を修正しました

yumiya-g committed:
Add: 日報にコメントしたユーザーのユーザー情報を表示する処理を追加

- ユーザ名またはメールアドレスを表示
- 退会済みの場合、「退会済みユーザー」と表示
- コメント投稿日時を表示

yumiya-g committed:
Add: 日報に投稿者と投稿日時を表示しました

yumiya-g committed:
Add: 本の詳細ページに、コメントを付けられる処理を実装しました

- `@commentable`には、日報または本のIDから取得したデータを格納

yumiya-g committed:
Fix: コメントフォームをパーシャルにしました

- 本と日報に投稿する際のコメント欄を共通化するため
- 削除ボタンクリック時に、確認ダイアログを表示

yumiya-g committed:
Add: コメントに`edit`と`destroy`アクションを追加しました

yumiya-g committed:
Fix: `destroy`アクションを実装しました

- インスタンス変数@commentableへの代入を無くし、`set_commentable`メソッドの戻り値を使用

yumiya-g committed:
Add: 日英の辞書ファイルを修正しました

- 日本語では、投稿日時を表示フォーマットを変更

yumiya-g committed:
Fix: 日報とコメントに、i18n対応を行いました

yumiya-g committed:
Add: 日英の辞書ファイルを修正しました

yumiya-g committed:
Rename: 日報または本の詳細ページでコメントを出力するテンプレート名を修正しました

yumiya-g committed:
Fix: コメントの編集機能を実装しました

- コメント編集ページには、コメント投稿先ページ（本または日報）へのアンカーテキストリンクを表示
- コメント編集を完了後、コメント先ページへリダイレクトする

yumiya-g committed:
Fix： 日報一覧にgem「kaminari」でページネーションを実装しました

- ページネーションリンクを追加
- 本やユーザーと同様に、カードレイアウトで日報を表示

yumiya-g committed:
Fix: erblintとrubocopのチェックと修正対応をしました

- 不要なヘルパーファイルを削除
- 不要な空白や空行を削除
- ダブルクォーテーションをシングルクォーテーションに変更

yumiya-g committed:
Fix: 日報一覧ページに表示される「この日報を表示」の出力処理を変更しました

- 日報詳細ページでテキストリンクが表示されないように調整

yumiya-g committed:
Fix: コメント投稿日時のHTML出力を調整しました

yumiya-g committed:
Delete: 使用していないテストファイルを削除しました

betachelsea commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,59 @@
+    comment = Comment.new(
+      content: comment_params[:content],
+      user_id: current_user.id,
+      commentable_type: @commentable.class.name,
+      commentable_id: @commentable[:id]
```
[nits] 次のように指定すると、type, id の投入をそれぞれうまくやってくれます 💡 

```ruby
comment = Comment.new(
  content: comment_params[:content],
  user_id: curretn_user.id, # なんならここも user: current_user でOK
  commentable: @commentable
)
```

yumiya-g replied:
アドバイスいただきありがとうございました。
こちらについてですが、修正後に提出物ページにて質問させていただいております。そちらもよろしくお願いいたします🙏🏻

betachelsea commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,59 @@
+  end
+
+  def set_comment
+    @comment = Comment.find(params[:id])
```
[MUST] 「コメントの削除・更新は投稿者本人のみ」という仕様があると思いますが、現在次のような手順を踏むと不正に編集できてしまうようです。

(編集のケース)
1. 特定のコメントのidを把握する (例: id=5とする)
2. 1の投稿者以外のユーザとしてログインしている時、`/comments/5/edit` にアクセスする
3. 正常にアクセス、コメント更新ができてしまう

また、削除はview側で本人以外に削除ボタンが露出しないようにすることで禁止しているようですが、サーバ側では削除を許してしまっているので、次のようなケースで削除されてしまいます。
- viewでうっかり削除ボタンを表示させてしまった場合 ( 新しく削除ボタンを付け足したとき、投稿者以外に表示しない、という実装をうっかり忘れてしまった、など)
- curlなどで直接DELETEリクエストされたとき 
  - ※編集でも直接PUTリクエストされたら発生します

直接リクエストなどでも確実に削除を禁止できるよう、サーバサイド側でも不正を防ぐような実装をしておいてください 💡

yumiya-g replied:
コメント作成者以外が編集できないよう、ログインユーザーIDとコメント作成者IDをチェックする処理を加えました。

betachelsea commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,81 @@
+    respond_to do |format|
+      if @report.save
+        format.html { redirect_to report_url(@report), notice: t('controllers.common.notice_create', name: Report.model_name.human) }
+        format.json { render :show, status: :created, location: @report }
```
[MUST] 注意点につぎのようにあるので、jsonレスポンス周りは削除しておいてください 🙏  コメント表記でも json の示唆があるようなので、全体的に見直してみてください！

> scaffoldの自動生成機能を使う場合は、JSONレスポンスのように使う予定のない機能は削除した状態でコミットする

yumiya-g replied:
JSON周りのコードの削除を行いました。

betachelsea commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,81 @@
+  end
+
+  # PATCH/PUT /reports/1 or /reports/1.json
+  def update
```
[MUST] Reportについては edit, destroy で 投稿者自身かどうかの確認をしていますが、updateは未実施のようですね。updateもリクエストされると更新してしまう、という状況が問題なので、対策が必要です。

`set_report` のタイミングで、 `Report.find(params[:id])` となっていますが、素直に利用して良いのは show の時だけのようです。edit, update, destroy のときは、ログインユーザの持っているreports のなかから探すのをオススメします。

```ruby
current_user.reports.find(params[:id])
```

yumiya-g replied:
ログインユーザーが作成した日報のみを取得して、編集（または削除）を実行する仕様に修正を行いました。

betachelsea commented on app/views/comments/_comment_lists.html.erb:
```diff
@@ -0,0 +1,26 @@
+<h2><%= t('views.common.title_index', name: Comment.model_name.human) %></h2>
+  <%= t("#{Comment.human_attribute_name('no_comments')}") if @comments.blank? %>
+  <% @comments.each do |comment| %>
+  <div>
```
[nits] 処理がネストしているところは、インデントが適用されていると読む時に助かります！ 🙏 

```ruby
<% @comments.each do |comment| %>
  <div>
    # ... 
```

とか

```ruby
<% if comment.user %>
  <%= comment.user.name.presence || comment.user.email %>
<% else %>
  <%= User.human_attribute_name('unsubscribed_user') %>
<% end %>
```

など

yumiya-g replied:
インデントが適切になっておらず失礼いたしました😓インデント調整を行いました。

betachelsea commented on app/views/reports/_report.html.erb:
```diff
@@ -0,0 +1,14 @@
+      <%= report.title %>
+    </h2>
+
+    <p><%= Comment.human_attribute_name('posted_by') %>:<%= report.user.name.presence || report.user.email %><br>
```
[nits] ユーザの名前 or email、を使うシーンは複数ありそうですね。何回か呼ぶ処理はモデルのほうにメソッドを定義してあげて、それを呼び出すとスマートかなと思います 💡

yumiya-g replied:
`Report`モデルに`account_name`メソッドを作成して、ユーザー名またはメールアドレスを表示する処理にしました。

betachelsea commented on config/routes.rb:
```diff
@@ -1,7 +1,17 @@
 Rails.application.routes.draw do
+  resources :reports
```
[nits] `reports` のルーティングは下記に `resources :reports do ~ end` で定義があるので、この行は消しても大丈夫そうです。

yumiya-g replied:
重複していた`resources :reports`を削除しました。

betachelsea commented on config/routes.rb:
```diff
@@ -1,7 +1,17 @@
   resources :users, only: %i(index show)
+
+  resources :books do
+    resources :comments, only: [:create, :edit, :destroy]
```
[nits] `edit` `destroy` は、下部の ` resources :comments, only: [:edit, :update, :destroy]` の指定と被るので、MECEを意識するなら `only: [:create]` だけにしておくと良さそうです 💡

yumiya-g replied:
こちらも重複していたので、`resources :comments, only: [:create]`としました。

betachelsea replied:
[nits] booksの comments については、まだ `only: [:create]` が未適用のようです〜 :eyes: コミット漏れでしょうか？

betachelsea commented on db/migrate/20231115132701_add_user_id_to_reports.rb:
```diff
@@ -0,0 +1,5 @@
+class AddUserIdToReports < ActiveRecord::Migration[7.0]
+  def change
+    add_column :reports, :user_id, :intger
```
[MUST] reportsのテーブルを作るmigrationは同じPullRequest内に定義してあるので、まとめてしまいたいですね。migrationはmainブランチに入って本番環境に適用したあとの修正は御法度ですが、今回のようにfeatureブランチ上で修正するのはありうる操作かなと思います。開発環境でロールバックを実行して、Reportに関するmigrationファイルを
ひとまとまりにする、という作業をやってみてください 💪 

(あと、 `:intger` ってtypoしているように思うのですが、動きましたね… 👀 )

参考: [Active Record マイグレーション \- Railsガイド - 4.1 ロールバック](https://railsguides.jp/active_record_migrations.html#%E3%83%AD%E3%83%BC%E3%83%AB%E3%83%90%E3%83%83%E3%82%AF)

yumiya-g replied:
ロールバック後、`20231114133218_create_reports.rb` に`user_id`を追加し、このファイルは削除しました。（typoには気づいてませんでした😓）

betachelsea commented on app/views/comments/_comment_lists.html.erb:
```diff
@@ -0,0 +1,26 @@
+    <br><%= l comment.created_at %></p>
+
+    <p><%= comment.content %></p>
+    <% if comment.user.present? && comment.user[:id] == current_user[:id] %>
```
[nits] `comment.user[:id]`  `current_user[:id]`という書き方ですが、`comment.user.id`  `current_user.id`と書くのがRailsらしいです 💡

yumiya-g replied:
ご指摘いただきありがとうございます。こちらも書き方を修正をしました。

betachelsea reviewed (commented):
コメントしました〜 ✍️

yumiya-g committed:
Fix: 他ユーザーのコメントを編集・削除できないよう修正しました

- Commentインスタンスの生成処理を修正
- コメントの編集・削除前に、ログインユーザーIDとコメント投稿者のユーザIDを比較する処理を実装

yumiya-g committed:
Delete: 不要なjsonレスポンスの削除を行いました

- 既存のBookモデルからもjsonレスポンスを削除しました

yumiya-g committed:
Fix: コメント一覧ページの表示を調整しました

- if文およびHTMLタグのインデントを調整しました
- 編集・削除ボタンの表示条件にて、ユーザーIDを取得方法を変更しました

yumiya-g committed:
Fix: ユーザー名またはユーザーメールアドレスの表示条件を調整しました

- Reportモデルに`account_name`メソッドを定義し、`report.html.erb`で呼び出す
- 不要になった`before_action`を削除

yumiya-g committed:
Fix: 日報のルーティング設定を修正しました

- 重複しているreportsのルーティング（resourcesメソッド）を削除
- 重複しているcommentsのルーティングを修正
	- `only`オプションの指定から`edit`と`destroy`の重複を解消

yumiya-g committed:
Fix: 日報（Reports）のマイグレーションを修正しました

- あとから`user_id`カラムを追加したマイグレーションファイルを削除
- `reports`テーブル作成時のマイグレーションファイルに`user_id`カラムを追加

yumiya-g committed:
Fix: 他のユーザーが作成した日報を、編集・削除できないようにしました

- `set_current_users_report`メソッドより、ログインユーザーが作成した日報のみを取得

betachelsea commented on app/controllers/books_controller.rb:
```diff
@@ -26,10 +29,8 @@ def create
     respond_to do |format|
       if @book.save
         format.html { redirect_to book_url(@book), notice: t('controllers.common.notice_create', name: Book.model_name.human) }
-        format.json { render :show, status: :created, location: @book }
```
おお、Bookのほうのcontrollerにもjson設定が残っていたのですね 👀  削除ありがとうございます！ 

[FYI] 今回、新規作成された ReportsController のファイルを拝見して json 設定について問題に思ってコメントしたのですが、実際のプロジェクトでPullRequestの本筋 (今回でいうと「コメントをつけられるようにすること」)以外の修正に手を伸ばしていくとキリが無くなってしまうので、こういうリファクタリングをしたくなった場合、このPullRequestではあえて何もしない & するなら別PullRequest( 「不要なformat.jsonを削除する」…とか )で提出しなおす、という方法がスマートかなと思います。

例えばすでにかなり歴史のあるサービスのリポジトリの場合など、過去に見逃してしまって追加されているけどほんとは修正したい処理がたくさんあったりしそうです。そういう過去の問題を同時に修正してしまいたい！という気持ちは出てくることはありますが、関連箇所の修正を諸々入れると、レビューの負荷、リリース時の問題発生のリスクがぐっと上がってしまい、なかなかリリースできなくなる…といった問題が発生してしまいます。
PullRequestは可能な限り主題に沿った、コンパクトな変更にとどめることを意識して進めるのが良いですね 💡  
究極をいうと「たった1行の変更」のためにPullRequestを作っても良いくらいだと思います :eyes: 

(今回は「コメントをつけられるようにする」自体がそこそこでかい変更なので、`Files changed` タブの内容もボリューミーですが、やむをえないかなと。)

今回のBooksControllerのjsonの修正については、そのままにしておいていただいてもOKですが、レビュアーによっては上記(主題に沿っていない)の理由にて、変更を取り消すように指示されるケースがありうる、という点をお伝えしておきます 🙏

yumiya-g replied:
> PullRequestの本筋 (今回でいうと「コメントをつけられるようにすること」)以外の修正に手を伸ばしていくとキリが無くなってしまうので、こういうリファクタリングをしたくなった場合、このPullRequestではあえて何もしない & するなら別PullRequest( 「不要なformat.jsonを削除する」…とか )で提出しなおす、という方法がスマートかなと思います。

確かにプラクティスの内容と別の部分まで修正してしまうのは良くなかったですね😓
現場で起こり得る問題なども参考になりました🙏🏻アドバイスいただきありがとうございます😃

betachelsea commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,81 @@
+      user_id: current_user.id
+    )
+
+    respond_to do |format|
```
[MUST] format利用が1つのみ(今回は html のみ)の場合、respond_to 〜 end は使わなくて大丈夫です！

```ruby
if @report.save
  redirect_to report_url...
else
  render :new, ...
end
```

yumiya-g replied:
`create`,`update`,`destroy`の該当箇所を修正します🙏🏻

betachelsea commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,62 @@
+  end
+
+  def set_comment
+    @comment = Comment.find(params[:id])
```
[nits] Reportの処理で、`current_user.reports.find(params[:id])` を使ってログイン者の日報のみを取る手法がありましたが、それをcomment で使っても良いな〜と思いました 💡

yumiya-g replied:
Commentでも同様の処理を実装してみます✊🏻

betachelsea commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,81 @@
+
+  def set_current_users_report
+    @report = current_user.reports.find(params[:id])
+  rescue StandardError
+    redirect_to action: :index
+    flash[:alert] = '他のユーザーが作成した日報は編集できません'
```
[FYI] こちら、例外について丁寧にメッセージを返していますね！ ユーザーにとってとても優しい設計とは思うのですが、少し気になるのは、このメッセージを見る人はおそらく意図的に悪いことをしようとした人がほとんど、という点です 🤔  悪いことしようとした人に親切にメッセージを返してあげるのは大変なので、その場合は問答無用で「エラーです」の自動画面に任せる(例外を出たら出たままとして扱う。Railsに任せて何もしない)、というのも一手ということをお伝えしておきます。

上記の話とはまたちょっとズレるかもなのですが、下記URLの記事・動画がエラー処理の考え方として大変参考になるのでオススメです！

参考: 
- [Railsアプリケーションにおけるエラー処理（例外処理）の考え方 \#Ruby \- Qiita](https://qiita.com/jnchito/items/3ef95ea144ed15df3637)
- [Exceptional Rails by Shinichi Maeshima \- Kaigi on Rails 2023](https://kaigionrails.org/2023/talks/willnet/)

yumiya-g replied:
> 悪いことしようとした人に親切にメッセージを返してあげるのは大変なので、その場合は問答無用で「エラーです」の自動画面に任せる(例外を出たら出たままとして扱う。Railsに任せて何もしない)、というのも一手ということをお伝えしておきます

今回は単純な内容のコメントでしたが、エラーメッセージに余計な情報は不要ですね💡

リンク先も確認いたします👨🏻‍💻
アドバイスいただきありがとうございます🙏

betachelsea commented on app/views/comments/_comment_lists.html.erb:
```diff
@@ -0,0 +1,27 @@
+      <% else %>
+        <%= User.human_attribute_name('unsubscribed_user') %>
+      <% end %>
+      <br><%= l comment.created_at %></p>
```
[IMO] (すみません、この辺は個々人の好みかもですが…。)

6行目の `<p>` の後、14行でも `<p>` があって、おや、閉じタグはどこかな…と思ったのですが、12行目であったのですね :eyes: この辺は好みにもよるかもしれませんが、インデントを徹底するなら次のような感じになりそうと思ったので共有します 🙏 

```erb
<p>
  <% if comment.user %> <!-- ここも1段下げた -->
    <%= comment.user.name.presence || comment.user.email %>
  <% else %>
    <%= User.human_attribute_name('unsubscribed_user') %>
  <% end %>
  <br> <!-- これは悩んだけど単独でおいておく… -->
  <%= l comment.created_at %>
</p><!-- 開きタグ、閉じタグは同一レベルに揃えておくと追いかけやすいので助かります -->
```

```erb
<!-- これは全体が短くてわかりやすい(認知しやすい)ので、このままで良いと思いました！ -->
<p><%= comment.content %></p>
```

yumiya-g replied:
了解しました。こちらも修正します👨🏻‍💻

betachelsea commented on app/models/report.rb:
```diff
@@ -0,0 +1,10 @@
+  has_many :comments, as: :commentable, dependent: :destroy
+end
+
+def account_name(report_owner_id)
+  @account_name = User.find(report_owner_id).name.presence || User.find(report_owner_id).email
+end
```
[MUST] Rubyは自由度がかなり高いので、`class Report < ApplicationRecord 〜 end` の外側に定義したものを呼び出せてしまうのですが、モデルのファイル ( `app/models/` の配下にあるファイル群 ) に Report クラス以外のものを書いておくのはかなり「レールから逸脱したコード」なので、NGかなと思いました 😢  

こういった、「どこにも属さないけどviewで使いたいお助けメソッド」は、helperを書くのが良いです。 helperは`app/helpers` 以下に書きます。「Rails カスタムHelper」とかで調べていただくと良さそうです。

ただ、今回のRailsのリポジトリの規模(小さめ)の場合、コードがあちこちに分散される(※)のも手間なので、おとなしく対象モデルのclassのメソッドとして定義するのが良いかと思います。上記のケースは、`class Report < ApplicationRecord 〜 end` の **外側** に定義されているので、classのメソッドとして成立していない点を注意してください。

※コードが増えてきてFatModelとかの問題が出てくると、decoratorなど用途によってコードの書き先を変えるような対応をしたりしますが、かなりややこしくなってくるし各社・各チームによって対応方針が違ったりするので、今のところ「へ〜そういうケースもあるんだ〜」くらいに思っていただければよいです 🙏   現時点ではModelファイルの正しい活用に注力してもらいたいです！ 💪 

----

改めて修正方針ですが、ソースコードをよく見ると、`User.find` が2回登場しており、ReportというよりUserに興味の高い処理のように思います。 `app/models/user.rb` にて、Userクラス内にメソッドを定義してみてください。また、その時Reportのオーナー情報は不要です。なぜなら次のように呼び出せばOKだからです！ 👍 

```ruby
report.user.account_name  #=> nameもしくはemailが得られる
```

( `app/helpers` にすでにそれらしいメソッドの定義があるようですが、練習と思って、改めてメソッド定義をやってみてください！ 💡  )

yumiya-g replied:
> ただ、今回のRailsのリポジトリの規模(小さめ)の場合、コードがあちこちに分散される(※)のも手間なので、おとなしく対象モデルのclassのメソッドとして定義するのが良いかと思います。上記のケースは、class Report < ApplicationRecord 〜 end の 外側 に定義されているので、classのメソッドとして成立していない点を注意してください。

ありがとうございました🙏
まず、そもそもですがReportクラスの中に定義すべきでした😅

ご指摘のとおりユーザー情報を取得する処理なので、Userクラスのメソッドとして再定義してみます。

betachelsea reviewed (commented):
拝見してコメントしました〜 ✍️   FBCコメントに書いていただいた質問には、FBCコメントの方で回答書きますね！

yumiya-g committed:
Fix: フォーマットがHTMLのみなので、espond_toメソッドを削除しました

 - 他ユーザーのコメントを編集した際に表示されるエラーメッセージも削除

yumiya-g committed:
Fix: <p>タグ内のインデントと閉じタグの位置を調整しました

yumiya-g committed:
Fix: 本のルーティング設定を修正しました

- 重複しているcommentsのルーティングを修正
  - onlyオプションの指定から`edit`と`destroy`の重複を解消

yumiya-g committed:
Fix: ユーザー名を表示するメソッド`account_name`をUserクラスのインスタンスメソッドに変更しました

- report.rbからコードを削除
- <p>タグの閉じタグの位置とインデントを調整

yumiya-g committed:
Fix: コメントを編集・削除する際に、ログインユーザーの作成したコメントのみを抽出する処理に変更しました

- エラーメッセージを削除
- 他ユーザーの編集した場合、そのコメントが関連付けられた本または日報の詳細ページへリダイレクト

betachelsea commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,73 @@
+
+  def set_current_users_report
+    @report = current_user.reports.find(params[:id])
+  rescue StandardError
+    redirect_to action: :index
```
[nits] 悪いことしようとした人へのメッセージは削除されましたね :eyes:  そうなると、「rescue して redirect してあげる」ことも不要かな〜と思います。ずっと後になって見直したとき、rescueしているところは特に何か重要な処理なのかな？と思われると、重点的なテストや動作確認などの保守コストが高まるので、不要なrescueはできるだけしない、と割り切った方がシンプルに保ててオススメです。(逆に考えると、必要な箇所は手厚く例外処理をしてあげたほうがいいですね！ )

yumiya-g replied:
> ずっと後になって見直したとき、rescueしているところは特に何か重要な処理なのかな？と思われると、重点的なテストや動作確認などの保守コストが高まるので、不要なrescueはできるだけしない、と割り切った方がシンプルに保ててオススメです。

ありがとうございます。
今回のように重要な処理ではない部分には`rescue`は不要とのことですので修正します🙏

betachelsea commented on app/models/user.rb:
```diff
@@ -7,4 +7,11 @@ class User < ApplicationRecord
+  has_many :reports, dependent: :destroy
+  has_many :comments, dependent: :nullify
+
+  def account_name(user)
+    [user.name, user.email].find(&:present?)
+  end
```
[MUST] userを引数としてもっていますが、Userモデルの中に生えているインスタンスメソッドなので、引数をやめて自分自身 ( self )を参照するようにしてみてください！ 

```ruby
# 例: 今回と全く関係ないが挨拶を出力するメソッドを生やす時

class User < ApplicationRecord
  # …中略…
  
  # #{self.name} ではなく #{name} でも問題なく動くはず。好みがありそう。
  # モデルクラスはattributesを参照するメソッドを持っているので、それを呼び出している。
  # @user.name を呼び出せるということは、自メソッド内でも `name` を呼べると考えて良い。
  def say_hello
    "私の名前は#{self.name}です。よろしくお願いします。"
  end
end
```

----

[MUST] また、処理内容が helper の中身そのままのようですので、この処理の内容を文章で説明してみてほしいです！ 🙏  → `[user.name, user.email].find(&:present?)`

yumiya-g replied:
ありがとうございます🙏

```ruby
report.user.class
#=> User(id: integer, email: string, encrypted_password: string, reset_password_token: string, reset_password_sent_at: datetime, remember_created_at: datetime, created_at: datetime, updated_at: datetime, name: string, postal_code: string, address: string, self_introduction: text)
```

より、`report.user`はUserクラスとなるため、
`report.user.account_name` はインスタンスメソッドとなり、`self`で自身（`report.user`）取得できるため、引数`user`は不要になる、との理解でよろしいでしょうか。

```ruby
def account_name
  [name, email].find(&:present?)
end
```

なお、rubocopでチェックを通したところ`[self.name, self.email]`は冗長とのことで、`self`を削除しました。

---
> 処理内容が helper の中身そのままのようですので、この処理の内容を文章で説明してみてほしいです！ 🙏

`[user.name, user.email].find(&:present?)`については、`[user.name, user.email].find( |u| u.present?)`を省略したもの。

`find`メソッド で配列の1番目から順にブロックパラメータ`u`に代入。`present?`メソッドにて`u`が`nil`かつ空でない場合`true`を返す処理なので、`user.name`が存在すればユーザー名を返し、ユーザ名がなければ（空なら）`user.email`でメールアドレスを返す処理となっています。

なお、下記のように`presence`メソッドと`||`で同じ処理を実現できます。

```ruby
def account_name
  name.presence || email
end
```


ですがこの場合、仮に3番目の条件が入った場合に`presence`メソッドを複数設定する必要が出てくるかと思い、`find`と`present?`メソッドでの処理を採用しました。

```ruby
def account_name
  name.presence || email.presence || address
end
```

betachelsea replied:
説明ありがとうございます！バッチリだと思いました 👍 
補足ですが、コードブロックの先頭に言語名を書くと、シンタックスハイライトされて見やすくなるので、ぜひ活用してください！ FBCのコメント欄でも同様に利用できます。

```
先頭をこう書くと…
```ruby
```

↓ 

```ruby
def account_name
  name.presence || email
end
```

yumiya-g replied:
ありがとうございます🙏シンタックスハイライトを追加しました👨🏻‍💻

betachelsea commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,73 @@
+  before_action :set_report, only: %i[show]
+  before_action :set_current_users_report, only: %i[edit update destroy]
+
+  # GET /reports or /reports.json
```
[nits] おっと、コメントに json の説明が残っていたようです… 👀

yumiya-g replied:
JSONに関する部分（`or /reports.json`など）を削除していきます。

betachelsea reviewed (commented):
修正お疲れ様です！ 🍵  細かいところは概ね問題ないと思いました！
モデルに生やした account_name について、追加で確認をお願いします 🙏  モデルの取り扱いはRailsを書いていく上で肝要なポイントなので、ぜひこの機会に色々な発見を得ていただきたいです！

yumiya-g committed:
Delete: コメントアウトから、JSONに関する説明を削除しました

yumiya-g committed:
Delete: 他ユーザーの日報を編集しようとした際の`resque`処理を削除しました

yumiya-g committed:
Fix: Userモデルのインスタンスメソッドに引数を渡さない処理に変更しました

betachelsea commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,60 @@
+  def set_comment
+    @commentable = Comment.find(params[:id]).commentable
+    @comment = current_user.comments.find(params[:id])
+  rescue StandardError
+    redirect_to @commentable
```
[nits] こちらも不要だと思いました〜 👀 
PullRequestのレビューでは、似たような箇所の指摘コメントは省略するケースがあるので、1件の修正については類似の箇所がないか見直すのをオススメします！ 💡

yumiya-g replied:
ご指摘いただきありがとうございます🙏
確認漏れでしたので修正します❗

betachelsea reviewed (commented):
修正お疲れ様でした！ 確認OKです〜 🙆 
※もし業務としてやるなら、[nits] の修正も終わるまでお願いしていたと思います！補足まで〜

yumiya-g committed:
Delete: 他ユーザーのコメントを編集しようとした際のresque処理を削除しました
